cosi.proto

```diff
@@ -49,58 +49,333 @@ extend google.protobuf.ServiceOptions {
+
+ message CreateBucketResponse {
+     // Bucket returned
+     Bucket bucket = 1;
+ }
+ 
+ message DeleteBucketRequest {
+     // The name of the bucket to be deleted.
+     // This field is REQUIRED.
+     string bucket_name = 1;
+    
+     // RequestProtocol, one of the predefined values
+     // Driver must check the protocol used to match 
+     // BucketClass:supportedProtocols - {"azureblob", "gcs", "s3", ... } [3]
+     // BucketRequest:protocol - use this as request protocol but check 
+     // if the protocol is in the BucketClass' suppportedProtocols
+     DataProtocol request_protocol = 2;
+
+     // DriverParameters, these are parameters that are extracted from 
+     // BuckerRequest and BucketClass so that the call has context.
+     // For example GCS require projectName for CreateBucket to succeed.
+     // BucketClass:provisioner - identify the  
+     // projectID if GCS
+     ProviderContext provider_context = 3; //- { azure_context, gcs_context, s3_context, generic_context}
+     map<string, string> driver_parameters = 4; //provider_context
+  
+     // Secrets required by driver to complete bucket deletion request.
+     // This field is OPTIONAL. Refer to the `Secrets Requirements`
+     // section on how to use this field.
+     // Azure:
+     //    message AuthenticationData {
+     //        option (cosi_secret) = true;
+     //        string StorageAccountName = 1;
+     //        string AccountKey = 2;
+     //        string SasToken = 3;
+     //    }
+     // GCS:
+     //    message AuthenticationData {
+     //        option (cosi_secret) = true;
+     //        string StorageAccountName = 1;
+     //        string PrivateKeyName = 2;
+     //        string PrivateKey = 3;
+     //    }
+     // S3:
+     //    message AuthenticationData {
+     //        option (cosi_secret) = true;
+     //        string AccessKeyId = 1;
+     //        string SecretKey = 2;
+     //        string StsToken = 3;
+     //        string UserName = 4;
+     //    }
+     map<string, string> secrets = 5;
+}
+
+
+
+ message DeleteBucketResponse {
+     // INTENTIONALLY BLANK
+ }
+
+// Error list for DeleteBucket
+// DELETE_BUCKET_DOESNOT_EXIST : Bucket specified does not exist 
+// DELETE_DELETE_INPROGRESS    : Delete bucket is in progress
+// DELETE_INVALID_CREDENTIALS  : resource deletion failed due to invalid credentials
+// DELETE_INTERNAL_ERROR       : Failed to execute the requested call
+// DELETE_INVALID_ARGUMENT     : validation of the input argument fails 
+
+
+
+ message S3Credentials {
+      string id = 1; // one of id, emailid, uri
+      string permission = 2;
+      string owner = 3;
+ }
+
+message GCSCredentials {
+     string entity = 1; //one ot userid, emailid, groupid, etc or 'allusers/allAuthenticatedUsers
+     string role = 2;
+     string domain = 3;
+     string project = 4;
+}
+
+message AzureCredentials {
+     string id = 1;
+     string permission = 2;
+}
+
+message GenericCredentials {
```

Comment:
Seems like we might be able to do generic credentials here too.  We need to see how credential information is handled in the api objects.